File: its-my-training-day/leetcode-top-75/questions/dynamic-programming/dp-1d/house-robber-ii/solution.cpp
#include "../../../../shared/cpp/ds.hpp"

#include <algorithm>
#include <cctype>
#include <cstdint>
#include <functional>
#include <string>
#include <unordered_map>
#include <vector>

int houseRobberII(const std::vector<int>& nums) {
    if (nums.empty()) return 0;
    if (nums.size() == 1) return nums[0];
    auto robRange = [&](int left, int right) {
        int prev1 = 0;
        int prev2 = 0;
        for (int i = left; i <= right; ++i) {
            int take = prev2 + nums[i];
            int skip = prev1;
            int current = take > skip ? take : skip;
            prev2 = prev1;
            prev1 = current;
        }
        return prev1;
    };
    int a = robRange(0, static_cast<int>(nums.size()) - 2);
    int b = robRange(1, static_cast<int>(nums.size()) - 1);
    return a > b ? a : b;
}

File: its-my-training-day/leetcode-top-75/questions/dynamic-programming/dp-1d/house-robber-ii/solution.py
import os
import sys
from typing import List, Dict, Optional, Tuple

ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../"))
if ROOT_DIR not in sys.path:
    sys.path.append(ROOT_DIR)

from shared.python.ds import ArrayList, ListNode, Stack, Queue, TreeNode, MinHeap, MaxHeap

def house_robber_ii(nums: List[int]) -> int:
    if not nums:
        return 0
    if len(nums) == 1:
        return nums[0]

    def rob_range(left: int, right: int) -> int:
        prev1 = 0
        prev2 = 0
        i = left
        while i <= right:
            take = prev2 + nums[i]
            skip = prev1
            current = take if take > skip else skip
            prev2 = prev1
            prev1 = current
            i += 1
        return prev1

    return max(rob_range(0, len(nums) - 2), rob_range(1, len(nums) - 1))
